### ISSUE TITLE
[question] Why does DataCache.GetAndChange not change the item in all possible cases?

### ISSUE THREAD

**Issue by ixje** [neo-project/neo#855]
I'm looking at 
https://github.com/neo-project/neo/blob/1e2c6c67795eed64c9266ec4292621f58a99e44f/neo/IO/Caching/DataCache.cs#L147

The name of the function suggests "You get the current item according to a key, and replace that with a new item". 

The following is a piece of code where we already found the item in the current cache. The above description matches the `if` branch but not the `else` branch of the code below.
https://github.com/neo-project/neo/blob/1e2c6c67795eed64c9266ec4292621f58a99e44f/neo/IO/Caching/DataCache.cs#L151-L162

Why are we not updating `trackable.Item` with `factory()` like the function name suggests? 

**To summarise what the function does**
We update `trackable.item` if:
- the item was found in the cache and its previous state was `DELETED`
- the item was not found in cache and not in the real DB

We do **not** update `trackable.item` if:
- the item was found in the cache, but it's previous state is `CHANGED` or `ADDED`
- the item was not found in cache, but found in the DB.

**Comment by ixje** [comment]
Additionally;
In all cases we update the `trackable.State` except when the item was found in cache, but its state was `ADDED` 😕

**Comment by ixje** [comment]
@erikzhang can you besides labelling also answer the question? You seem to be the author of that class. Thanks.

**Comment by shargon** [comment]
>You get the current item according to a key, and replace that with a new item".

No, You get the current item according to a key, and if not exists, you will create a new item like this

**Comment by ixje** [comment]
Uhh...what is "Change" in the function name supposed to refer to !? 😖 
Should this instead be named "GetOrCreate"?

Also, why do we update the state to `CHANGED` for an item that is found in cache and simply returned (line 161)? Getting an item should not imply modification, right?

**Comment by shargon** [comment]
Agree, should be better `GetOrCreate`

**Comment by erikzhang** [comment]
If you keep the `factory` as the default value `null`, it won't create anything.

**Comment by shargon** [comment]
> Also, why do we update the state to CHANGED for an item that is found in cache and simply returned (line 161)? Getting an item should not imply modification, right?

Is changed, because you will lost the control of this item, so you need to check as `CHANGED`

I will optimize this `CHANGED` if you don't change anything

**Comment by ixje** [comment]
> Is changed, because you will lost the control of this item, so you need to check as CHANGED

You mean because the value is returned by reference and potentially gets changed by the caller. Is that the right understanding?

If so, then we must do the same (update `State` to `CHANGED`) for `TryGet` at line 228
https://github.com/neo-project/neo/blob/1e2c6c67795eed64c9266ec4292621f58a99e44f/neo/IO/Caching/DataCache.cs#L221-L228

----

> If you keep the factory as the default value null, it won't create anything.

I understand. In that case you should thus actually call `TryGet` instead and test for `null`, because by not providing a factory you basically indicate you only want to get a value. It seems like this `GetAndChange` function is trying to do too much, would you agree? There's `Add`, `TryGet` and `GetOrAdd` that seem to cover the same functionality.

**Comment by shargon** [comment]
Then, that is the reason for the name GetAndChange, because with the other you are not able to alter the entry

**Comment by ixje** [comment]
> Then, that is the reason for the name GetAndChange, because with the other you are not able to alter the entry

Can you be specific what reason you're talking about ? 

`TryGet()` internally uses the same logic (returning the result of `TryGetInternal()`) as `GetAndChange()`, so I see no reason why would be able to alter and the other not. 

`GetAndChange` duplicates the same logic as TryGet/GetOrAdd and Add, except because it doesn't internally use those functions you get these edge cases where this supposed convenience function is updated whereas the others are not (as I pointed out) above leading to differences
> If so, then we must do the same (update State to CHANGED) for TryGet at line 228

**Comment by shargon** [comment]
Because anything that you do with `TryGet` never will be committed, with `GetAndChange`, your changes will be applied

**Comment by ixje** [comment]
> Because anything that you do with `TryGet` never will be committed, with `GetAndChange`, your changes will be applied

This modifies `0xAA` to `0xBB` just fine with `TryGet()`
```C#
        static void Main(string[] args)
        {
            LevelDBStore LDBStore = new LevelDBStore("/tmp/tmpchain/");
            UInt160 script_hash = new UInt160();
            byte[] key = new byte[] {0xBE, 0xEF};
            
            StorageKey my_key = new StorageKey
            {
                ScriptHash =  script_hash,
                Key = key
            };

            using (Snapshot snapshot = LDBStore.GetSnapshot())
            {
                snapshot.Storages.Add(my_key, new StorageItem {Value = new byte[] {0xAA}, IsConstant = false});
                StorageItem si = snapshot.Storages.TryGet(my_key);
                si.Value = new byte[] { 0xBB };

                snapshot.Commit();
            }
            
            LDBStore.Dispose();
            
            // re-open to ensure we didn't cache anything
            LDBStore = new LevelDBStore("/tmp/tmpchain");
            using (Snapshot snapshot = LDBStore.GetSnapshot())
            {
                StorageItem si = snapshot.Storages.TryGet(my_key);
                Console.WriteLine($"{BitConverter.ToString(si.Value)}"); // prints BB
            }
```

If we `TryGet()` and item that was not in cache but is in the real DB, then modifying doesn't work because of
https://github.com/neo-project/neo/blob/1e2c6c67795eed64c9266ec4292621f58a99e44f/neo/IO/Caching/DataCache.cs#L236
However, change that to `TrackeState.Changed` like the `GetAndChange()` does here for the same situation https://github.com/neo-project/neo/blob/1e2c6c67795eed64c9266ec4292621f58a99e44f/neo/IO/Caching/DataCache.cs#L179
and you now have the same capabilities.

**Comment by shargon** [comment]
``` snapshot.Storages.Add(my_key, new StorageItem {Value = new byte[] {0xAA}, IsConstant = false});
                StorageItem si = snapshot.Storages.TryGet(my_key);
```
Use different snapshots, if you add a new key, is changed and cached, so when you get again, you will get the same object.
If you use different snapshot for get and for Add, you will get different results.

**Comment by ixje** [comment]
Uh yes, that's what I said in the last part of my post.

**Comment by shargon** [comment]
Yes, my fault, i started to answer when reading the code, sorry...

**Comment by lock9** [comment]
Hi @shargon  and @ixje, any progress in this discussion?

**Comment by lock9** [comment]
Hi guys!
Since there was no activity in the last 11 days, I'm going to close this.
@ixje if you think this was a mistake, please contact me in Discord / Slack so I can reopen it for you.
Thanks.

**Comment by vncoelho** [comment]
I think that we should leave this issue open until we document the code properly and/or rename some functions.
